src/controllers/aer_controller.hpp

```diff
@@ -887,20 +949,61 @@ Result Controller::execute(std::vector<Circuit> &circuits,
 #endif
     }
 #endif
+
     // then- and else-blocks have intentionally duplication.
     // Nested omp has significant overheads even though a guard condition
     // exists.
     const int NUM_RESULTS = result.results.size();
-    if (parallel_experiments_ > 1) {
-#pragma omp parallel for num_threads(parallel_experiments_)
+    if (parallel_experiments_ > 1 && sim_device_ != Device::GPU && !multi_chunk_req) {
+#pragma omp parallel for if (parallel_experiments_ > 1) num_threads(parallel_experiments_)
       for (int j = 0; j < result.results.size(); ++j) {
-        execute_circuit(circuits[j], noise_model, methods[j],
-                        config, result.results[j]);
+        run_circuit(circuits[j], noise_model,methods[j],
+                    config, circuits[j].shots, circuits[j].seed,
+                    result.results[j],multi_chunk[j]);
       }
-    } else {
-      for (int j = 0; j < result.results.size(); ++j) {
-        execute_circuit(circuits[j], noise_model, methods[j],
-                        config, result.results[j]);
+    }
+    else{
+      bool batch_enable = false;
+      if(batched_shots_optimization_ && batched_shots_optimization_threshold_ >= max_qubits_){
+        if(one_method && sim_device_ == Device::GPU && !multi_chunk_req && circuits.size() > num_gpus_ && 
+                max_batched_states_ > 1 && max_batched_states_ >= num_gpus_){
+          batch_enable = true;
+          for (size_t j = 0; j < circuits.size(); j++) {
+            if(methods[j] != Method::statevector && methods[j] != Method::density_matrix){
+              batch_enable = false;
+              break;
+            }
+            if(noise_model.has_quantum_errors() && methods[j] == Method::statevector){
+              batch_enable = false;
+              break;
+            }
+            if(!check_measure_sampling_opt(circuits[j], methods[j])){
+              batch_enable = false;
+              break;
+            }
+          }
+        }
+      }
```

Comment:
Can we move the above change to `set_parallelization_experiments()`? (In that case `batch_enable` will be a class field.)

Reply:
moved to `set_parallelization_experiments()`